docs: add Pallas-Vesta curves and Simplified SWU documentation (#850)
zcash/halo2: book/src/background/curves.md

![](https://i.imgur.com/bNMyMRu.png)

### Pallas-Vesta curves

> The Pallas and Vesta curves form a 2-cycle of elliptic curves designed specifically for Halo 2. They are defined over finite fields with highly 2-adic structure, meaning they have large multiplicative subgroups of order $2^S$ where $S = 32$.

Reviewer: ```suggestion
The Pallas and Vesta curves form a 2-cycle of elliptic curves designed to allow efficient recursion in discrete-log-based proof systems such as Halo 2. They are defined over finite fields with highly 2-adic structure, meaning they have large multiplicative subgroups of order $2^S$, where in this case $S = 32$.
```

**Pallas curve ($E_p/\mathbb{F}_p$):**
- Base field: $\mathbb{F}_p$ where $p = 2^{254} + t_p$ and $t_p = 45560315531419706090280762371685220353$
- Scalar field: $\mathbb{F}_q$ where $q = 2^{254} + t_q$ and $t_q = 45560315531506369815346746415080538113$

> - Curve equation: $y^2 = x^3 + 5$

Reviewer: I checked that the values of $t_p$ and $t_q$ match https://github.com/zcash/pasta (and I know everything else does).

[cfrg-hash-to-curve]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/?include_text=1

### Simplified SWU

The Simplified SWU (Shallue-van de Woestijne-Ulas) method is an efficient hash-to-curve algorithm that maps field elements to curve points. It is particularly well-suited for curves of the form $y^2 = x^3 + b$ where $b \neq 0$.

> For a field element $u \in \mathbb{F}_p$, the Simplified SWU algorithm works as follows:

Reviewer: This matches the core of the algorithm in the paper, but it doesn't mention that for curves with j-invariant 0 like Pallas and Vesta, you need to do the hash-to-curve operation for an isogenous curve and then map to the target curve using the isogeny. There are additional complications as noted in my second suggestion below.

It probably isn't useful to spell out all the details here. Instead we should point to the [Sage reference implementation](htttps://github.com/zcash/pasta/blob/master/hashtocurve.sage) (which is algorithmically optimized, including some changes needed to support constant-time operation in a real implementation*, but well commented).

*Obviously Sage itself can't support constant-time operation.

```suggestion
For a field element $u \in \mathbb{F}_p$, the core of the Simplified SWU algorithm works as follows:
```

1. **Precomputation**: Compute $Z = -b/A$ where $A$ is a non-square in $\mathbb{F}_p$
2. **Mapping**: For input $u$, compute:
   - $t_1 = -Z \cdot (1 + u^2)$
   - $t_2 = -Z \cdot (1 - u^2)$
   - $x_1 = t_1^3 + b$
   - $x_2 = t_2^3 + b$
   - $x_3 = Z \cdot (t_1^2 \cdot t_2^2)$

3. **Selection**: Choose the first $x_i$ that is a square in $\mathbb{F}_p$
4. **Square root**: Compute $y = \sqrt{x_i^3 + b}$
5. **Sign adjustment**: If $u \cdot y$ is negative, negate $y$

The result is the curve point $(x_i, y)$.

> 

Reviewer: ```suggestion

In practice for curves with $j$-invariant $0$ such as Pallas and Vesta, this algorithm cannot be used as-is. Instead we map $u$ to an isogenous curve with a non-zero $j$-invariant ("iso-Pallas" or "iso-Vesta"), and then apply the isogeny to give a point on the target curve. Also, the full hash-to-curve operation involves pre-hashing the input with domain separation, and adding the results of two such mappings to the target curve (or equivalently, adding them on the isogenous curve and then applying the isogeny). For full details see the [Sage reference implementation](https://github.com/zcash/pasta/blob/master/hashtocurve.sage).

```
